# Читання даних з веб-сторінок

---

## 0. Інсталяція потрібних пакетів

Для початку варто інсталювати пакети `rvest`, `dplyr`, `stringr`.

In [1]:
# install.packages("rvest") # парсер веб-сторінок
# install.packages("dplyr") # маніпулювання та вибірка даних
# install.packages("stringr") # робота з текстом/рядками

Завантажуємо пакет у память, щоб можна було використовувати функції.

In [3]:
library(rvest)
library(dplyr)
library(stringr)

## 1. Читаємо вміст вебсторінки та потрібні дані

Вказуємо адресу ресурсу для читання даних:

In [4]:
url <- "https://abit-poisk.org.ua/rate2022/direction/1032336"
#url <- "https://abit-poisk.org.ua/rate2022/direction/984185"
url

[1] "https://abit-poisk.org.ua/rate2022/direction/1032336"

Читаємо сторінку у змінну у форматі `html` як набір тексту:

In [5]:
page <- read_html(url)
page

{html_document}
<html lang="uk">
[1] <head>\n<meta http-equiv="Content-Type" content="text/html; charset=UTF-8 ...
[2] <body>\n<div id="js-ba-options" data-options='{"banners":[["\\/assets\\/i ...

Витягуємо з інформації про сторінку усі потрібні нам теги, у даному випадку це таблиці, тег `table`.

In [6]:
tables <- html_nodes(page, "table")
tables

{xml_nodeset (2)}
[1] <table class="table table-bordered table-hover">\n<thead><tr>\n<th class= ...
[2] <table class="table table-bordered table-hover" style="padding-right: 0;p ...

Конвертуємо отриману таблицю з формату html у `data.frame` та переглядаємо перших кілька рядків таблиці, щоб переконатися, що усе `ОК`.

In [12]:
data <- as.data.frame(html_table(tables[1]), fill = T)
data |> head()

X.
1 1 
2 2 
3 3 
4 4 
5 5 
6 6 
  ПІБ                                                                                                                                                                                                                       
1 1\n                                                        \n                                120-7906910  В. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 3
2 2\n                                                        \n                                Мініч К. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1       
3 3\n                                                        \n                                Редколіс А. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1    
4 4\n                                                        \n                                Скулкін О. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1     
5 5\n                                                        \n                                Мартинчук О. П.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1   
6 6\n                                                        \n                                Божок М. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1       
  П Заг.бал Статус       
1 3 160.95  До наказу (б)
2 1 197.90  До наказу (б)
3 1 194.00  До наказу (б)
4 1 180.70  До наказу (к)
5 1 179.50  До наказу (к)
6 1 177.40  До наказу (к)
  Складові.заг..балу                                                                                                                                                                                                                                                                                                                                          
1 160  168  151                                \n                                \n                                    Українська мова: 160\n\n                                                                                    Математика: 168\n\n                                                                                    Історія України: 151
2 194  200  200                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 200
3 200  185  200                                \n                                \n                                    Українська мова: 200\n\n                                                                                    Математика: 185\n\n                                                                                    Історія України: 200
4 194  172  176                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 172\n\n                                                                                    Історія України: 176
5 172  192  170                                \n                                \n                                    Українська мова: 172\n\n                                                                                    Математика: 192\n\n                                                                                    Історія України: 170
6 174  185  170                                \n                                \n             

## 2. Очистка та трансформація даних

### 2.1. Видалення зайвих стовпців

Видалимо 3-й та 7-й стовпець з інформацією про квоту

In [13]:
data |> select(c(3,7)) |> head()

П Квота  
1 3 Квота-1
2 1 —      
3 1 —      
4 1 —      
5 1 —      
6 1 —

In [14]:
data <- data |> select(-c(3,7))
data |> head()

X.
1 1 
2 2 
3 3 
4 4 
5 5 
6 6 
  ПІБ                                                                                                                                                                                                                       
1 1\n                                                        \n                                120-7906910  В. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 3
2 2\n                                                        \n                                Мініч К. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1       
3 3\n                                                        \n                                Редколіс А. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1    
4 4\n                                                        \n                                Скулкін О. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1     
5 5\n                                                        \n                                Мартинчук О. П.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1   
6 6\n                                                        \n                                Божок М. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1       
  Заг.бал Статус       
1 160.95  До наказу (б)
2 197.90  До наказу (б)
3 194.00  До наказу (б)
4 180.70  До наказу (к)
5 179.50  До наказу (к)
6 177.40  До наказу (к)
  Складові.заг..балу                                                                                                                                                                                                                                                                                                                                          
1 160  168  151                                \n                                \n                                    Українська мова: 160\n\n                                                                                    Математика: 168\n\n                                                                                    Історія України: 151
2 194  200  200                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 200
3 200  185  200                                \n                                \n                                    Українська мова: 200\n\n                                                                                    Математика: 185\n\n                                                                                    Історія України: 200
4 194  172  176                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 172\n\n                                                                                    Історія України: 176
5 172  192  170                                \n                                \n                                    Українська мова: 172\n\n                                                                                    Математика: 192\n\n                                                                                    Історія України: 170
6 174  185  170                                \n                                \n                           

### 2.2. Іменування стовпців

Замінимо назви стовпців дата-фрейму для зручності використання під час роботи з даними

In [15]:
colnames(data) <- c("Rank", "Name", "TotalGrade", "StudyType", "Grades", "IsStudent")
data |> head()

Rank
1 1   
2 2   
3 3   
4 4   
5 5   
6 6   
  Name                                                                                                                                                                                                                      
1 1\n                                                        \n                                120-7906910  В. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 3
2 2\n                                                        \n                                Мініч К. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1       
3 3\n                                                        \n                                Редколіс А. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1    
4 4\n                                                        \n                                Скулкін О. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1     
5 5\n                                                        \n                                Мартинчук О. П.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1   
6 6\n                                                        \n                                Божок М. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1       
  TotalGrade StudyType    
1 160.95     До наказу (б)
2 197.90     До наказу (б)
3 194.00     До наказу (б)
4 180.70     До наказу (к)
5 179.50     До наказу (к)
6 177.40     До наказу (к)
  Grades                                                                                                                                                                                                                                                                                                                                                      
1 160  168  151                                \n                                \n                                    Українська мова: 160\n\n                                                                                    Математика: 168\n\n                                                                                    Історія України: 151
2 194  200  200                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 200
3 200  185  200                                \n                                \n                                    Українська мова: 200\n\n                                                                                    Математика: 185\n\n                                                                                    Історія України: 200
4 194  172  176                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 172\n\n                                                                                    Історія України: 176
5 172  192  170                                \n                                \n                                    Українська мова: 172\n\n                                                                                    Математика: 192\n\n                                                                                    Історія України: 170
6 174  185  170                                \n                          

### 2.3. Трансформація даних у стовпцях

Підготуємо наш набір даних до вигляду, коли у нас будуть стовпці у вигляді

|Rank|Name|StudyType|UkrGrade|MathGrade|HistoryGrade|TotalGrade|IsStudent|
|---|---|---|---|---|---|---|---|
|1|Прізвище І.І.|Бюджет|158|198|184|172.4|TRUE|

`Rank` лишаємо без змін.

#### 2.3.1. Очистка поля `Name`

Очистимо стовпець `Name`. Для початку переглянемо, що насправді може бути у цьому стовпці:

In [19]:
data$Name[5]

[1] "5\n                                                        \n                                Мартинчук О. П.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1"

Переглянемо скільки символів присутньо у цьому записі

In [17]:
nchar(data$Name)

[1] 213 206 209 208 210 206 207 202 207 206 210 209 204 205 208 208 201 212 211
[20] 211 205 203 205 207 209 205 209 207 208 209 209 205 208 210 213 206 208 205
[39] 209 204 205 211 207 212 208 204 204 209 206 211 210 207 208 207 204 204 214
[58] 210 210 209 208 209 205 208 208 207 211 204 209

З чого він складається? Цифра на початку, `\n` та `\t`, багато зайвих пробілів та інформація про пріоритет вкінці  `Пр.`. 

Для початку позбудемося `\n` та `\t`.

In [18]:
name <- data$Name[5]
name <- str_replace_all(name, "\n", "")
name

[1] "5                                                                                        Мартинчук О. П.                                                        \t\t\t\t\t\t                                  Пр. 1"

In [20]:
name <- str_replace_all(name, "\t", "")
name

[1] "5                                                                                        Мартинчук О. П.                                                                                          Пр. 1"

Тепер позбудемося пробілів, наприклад, замінимо 2 пробіли на 1.

In [21]:
str_replace_all(name, "  ", " ")

[1] "5                                            Мартинчук О. П.                                             Пр. 1"

Стало менше проте ми їх не позбулися (((

In [22]:
name <- str_replace_all(name, "  ", " ")
name <- str_replace_all(name, "  ", " ")
name

[1] "5                      Мартинчук О. П.                       Пр. 1"

Майже, але ще не все

In [23]:
name <- str_replace_all(name, "  ", " ")
name <- str_replace_all(name, "  ", " ")
name

[1] "5      Мартинчук О. П.      Пр. 1"

Майже вийшло, проте ми не можемо гарантувати, що усі дані мають таку ж кількість пробілів. Варто написати алгоритм, що буде видаляти усі пробіли за 1 раз.

In [24]:
text_length = 0
text = name

# доки довжина тексту змінюється
while(text_length != nchar(text)) {
    # рахуємо скільки символів у тексті
    text_length <- nchar(text)    
    # замінюємо 2 пробіли на 1
    text <- str_replace_all(text, "  ", " ")   
}
text

[1] "5 Мартинчук О. П. Пр. 1"

Далі загорнемо це все у функцію:

In [25]:
text_cleaner <- function(text) {
    
    text <- str_replace_all(text, "\t", "")
    text <- str_replace_all(text, "\n", "")
    
    text_length = 0

    while(text_length != nchar(text)) {
        text_length <- nchar(text)    
        text <- str_replace_all(text, "  ", " ")   
    }
    text
}

Перевіримо чи функція працює на 1 записі:

In [26]:
text_cleaner(data$Name[3])

[1] "3 Редколіс А. В. Пр. 1"

Перевіримо чи працює функція у парі з `sapply()` для усього ветора імен

In [32]:
std_names <- sapply(data$Name, text_cleaner)
names(std_names) <- NULL
std_names

[1] "1 120-7906910 В. О. Пр. 3" "2 Мініч К. В. Пр. 1"      
 [3] "3 Редколіс А. В. Пр. 1"    "4 Скулкін О. В. Пр. 1"    
 [5] "5 Мартинчук О. П. Пр. 1"   "6 Божок М. О. Пр. 1"      
 [7] "7 Шевчук К. В. Пр. 1"      "8 Новік А. І. К"          
 [9] "9 Пінчук О. О. Пр. 5"      "10 Шуль М. В. Пр. 1"      
[11] "11 Мулярчук М. Д. Пр. 1"   "12 Климець М. П. Пр. 3"   
[13] "13 Зозуля О. В. К"         "14 Швецова Ю. С. К"       
[15] "15 Черняк К. В. Пр. 3"     "16 Черниш О. В. Пр. 4"    
[17] "17 Чих О. В. К"            "18 Харитонова І. С. Пр. 1"
[19] "19 Кучинська Ю. В. Пр. 4"  "20 Требухова В. І. Пр. 1"

Як видалити `#Номер` на початку та `Пр. Х` вкінці?

In [21]:
#indexes <- as.data.frame(str_locate(std_names, "Пр."))
#head(indexes, 10)

Варто відмітити, що присутні пропуски у таблиці, а це означає, що `Пр.` не було знайдено у записі. І якщо ми зробимо просто видалення за індексом `Пр.`, то окремі записи будуть знищені. Зверніть увагу на запис `#10`.

In [22]:
#names <- str_sub(std_names, start = 1L, end = indexes$start-2)
#names

Напишемо складніший алгоритм з перевіркою наявності `Пр.` у рядку:

In [27]:
#std_names <- ifelse(is.na(str_sub(std_names, start = 1L, end = indexes$start-2)),
#                    std_names, 
#                    str_sub(std_names, start = 1L, end = indexes$start-2))
#std_names

In [29]:
data <- data[1:20,]

In [30]:
data

Rank
1   1  
2   2  
3   3  
4   4  
5   5  
6   6  
7   7  
8   8  
9   9  
10 10  
11 11  
12 12  
13 13  
14 14  
15 15  
16 16  
17 17  
18 18  
19 19  
20 20  
   Name                                                                                                                                                                                                                      
1  1\n                                                        \n                                120-7906910  В. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 3
2  2\n                                                        \n                                Мініч К. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1       
3  3\n                                                        \n                                Редколіс А. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1    
4  4\n                                                        \n                                Скулкін О. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1     
5  5\n                                                        \n                                Мартинчук О. П.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1   
6  6\n                                                        \n                                Божок М. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1       
7  7\n                                                        \n                                Шевчук К. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1      
8  8\n                                                        \n                                Новік А. І.                            \n\n                            \n\t\t\t\t\t\t                                  К           
9  9\n                                                        \n                                Пінчук О. О.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 5      
10 10\n                                                        \n                                Шуль М. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1       
11 11\n                                                        \n                                Мулярчук М. Д.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 1   
12 12\n                                                        \n                                Климець М. П.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 3    
13 13\n                                                        \n                                Зозуля О. В.                            \n\n                            \n\t\t\t\t\t\t                                  К         
14 14\n                                                        \n                                Швецова Ю. С.                            \n\n                            \n\t\t\t\t\t\t                                  К        
15 15\n                                                        \n                                Черняк К. В.                            \n\n                            \n\t\t\t\t\t\t                                  Пр. 3     
16 16\n                                                        \n                                Черниш О. В.                            \n\n                            \n\t\t\t\t\t\t          

Далі нам потрібно видалити цифру (порядковий номер) на початку рядка. Алгоритм наступний: розбиваємо на частини рядка через пробіл, а потім усі крім першого "склеюємо" докупи.

In [33]:
x <- unlist(str_split(std_names[4], ' '))
x

[1] "4"       "Скулкін" "О."      "В."      "Пр."     "1"

In [34]:
x[2:4]

[1] "Скулкін" "О."      "В."

In [35]:
paste(x[2:4], collapse = " ")

[1] "Скулкін О. В."

Опишемо це як окрему функцію:

In [27]:
#remove_digit <- function(name) {
#    name <- unlist(str_split(name, ' '))
#    paste(name[2:length(name)], collapse = " ")
#}
#remove_digit(std_names[2])

Щоб виконати цю функцію для усього списку використаємо `lapply()`:

In [28]:
#std_names <- lapply(std_names, remove_digit)
#std_names |> head()

Тепер зберемо докупи усі етапи очистки поля Name в 1 функцію.

In [36]:
#clean_names <- function(names) {
#    student_names <- sapply(names, text_cleaner)
#    names(student_names) <- NULL
#    student_names <- ifelse(is.na(str_sub(student_names, start = 1L, end = indexes$start-2)), 
#                           student_names, str_sub(student_names, start = 1L, end = indexes$start-2))
#    student_names <- lapply(student_names, remove_digit)
#    return(student_names)
#}

get_name_only <- function(name){
    x <- unlist(str_split(name, " "))[2:4]
    paste(x, collapse = " ")
}

clean_names <- function(names) {
    student_names <- sapply(names, text_cleaner)
    names(student_names) <- NULL
    student_names <- lapply(student_names, get_name_only)
    student_names 
}


In [37]:
clean_names(data$Name) |> head()

[[1]]
[1] "120-7906910 В. О."

[[2]]
[1] "Мініч К. В."

[[3]]
[1] "Редколіс А. В."

[[4]]
[1] "Скулкін О. В."

[[5]]
[1] "Мартинчук О. П."

[[6]]
[1] "Божок М. О."

Замінимо дані у дата-фреймі:

In [38]:
data <- data |>
    mutate(Name = unlist(clean_names(Name)))

In [39]:
data |> head()

Rank Name              TotalGrade StudyType    
1 1    120-7906910 В. О. 160.95     До наказу (б)
2 2    Мініч К. В.       197.90     До наказу (б)
3 3    Редколіс А. В.    194.00     До наказу (б)
4 4    Скулкін О. В.     180.70     До наказу (к)
5 5    Мартинчук О. П.   179.50     До наказу (к)
6 6    Божок М. О.       177.40     До наказу (к)
  Grades                                                                                                                                                                                                                                                                                                                                                      
1 160  168  151                                \n                                \n                                    Українська мова: 160\n\n                                                                                    Математика: 168\n\n                                                                                    Історія України: 151
2 194  200  200                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 200
3 200  185  200                                \n                                \n                                    Українська мова: 200\n\n                                                                                    Математика: 185\n\n                                                                                    Історія України: 200
4 194  172  176                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 172\n\n                                                                                    Історія України: 176
5 172  192  170                                \n                                \n                                    Українська мова: 172\n\n                                                                                    Математика: 192\n\n                                                                                    Історія України: 170
6 174  185  170                                \n                                \n                                    Українська мова: 174\n\n                                                                                    Математика: 185\n\n                                                                                    Історія України: 170
  IsStudent
1 +        
2 +        
3 +        
4 +        
5 +        
6 +

#### 2.3.2. Робота з полем `StudyType`

In [40]:
unique(data$StudyType)

[1] "До наказу (б)" "До наказу (к)" "Допущено"

In [41]:
as.vector(data$StudyType)

[1] "До наказу (б)" "До наказу (б)" "До наказу (б)" "До наказу (к)"
 [5] "До наказу (к)" "До наказу (к)" "До наказу (к)" "До наказу (к)"
 [9] "До наказу (к)" "До наказу (к)" "До наказу (к)" "До наказу (к)"
[13] "До наказу (к)" "До наказу (к)" "До наказу (к)" "Допущено"     
[17] "Допущено"      "Допущено"      "Допущено"      "Допущено"

In [42]:
data <- data |>
    mutate(StudyType = ifelse(StudyType == "До наказу (б)", "Бюджет",
                              ifelse(StudyType == "До наказу (к)", "Контракт",
                                     "Не навчається"))       
)

In [43]:
data |> head()

Rank Name              TotalGrade StudyType
1 1    120-7906910 В. О. 160.95     Бюджет   
2 2    Мініч К. В.       197.90     Бюджет   
3 3    Редколіс А. В.    194.00     Бюджет   
4 4    Скулкін О. В.     180.70     Контракт 
5 5    Мартинчук О. П.   179.50     Контракт 
6 6    Божок М. О.       177.40     Контракт 
  Grades                                                                                                                                                                                                                                                                                                                                                      
1 160  168  151                                \n                                \n                                    Українська мова: 160\n\n                                                                                    Математика: 168\n\n                                                                                    Історія України: 151
2 194  200  200                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 200
3 200  185  200                                \n                                \n                                    Українська мова: 200\n\n                                                                                    Математика: 185\n\n                                                                                    Історія України: 200
4 194  172  176                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 172\n\n                                                                                    Історія України: 176
5 172  192  170                                \n                                \n                                    Українська мова: 172\n\n                                                                                    Математика: 192\n\n                                                                                    Історія України: 170
6 174  185  170                                \n                                \n                                    Українська мова: 174\n\n                                                                                    Математика: 185\n\n                                                                                    Історія України: 170
  IsStudent
1 +        
2 +        
3 +        
4 +        
5 +        
6 +

#### 2.3.3. Очистка інформації про оцінки (`Grades`)

Як бачимо, поле `Grades` містить інформацію про усі оцінки + нави предметів. Наша задача - розділити ці дані на 3 окремі стовпці з інформацією лише про оцінки, а назви предметів варто закодувати у назвах стовпців.

In [44]:
data |> select(Grades) |> head(3)

Grades                                                                                                                                                                                                                                                                                                                                                      
1 160  168  151                                \n                                \n                                    Українська мова: 160\n\n                                                                                    Математика: 168\n\n                                                                                    Історія України: 151
2 194  200  200                                \n                                \n                                    Українська мова: 194\n\n                                                                                    Математика: 200\n\n                                                                                    Історія України: 200
3 200  185  200                                \n                                \n                                    Українська мова: 200\n\n                                                                                    Математика: 185\n\n                                                                                    Історія України: 200

In [45]:
data$Grades[1]

[1] "160  168  151                                \n                                \n                                    Українська мова: 160\n\n                                                                                    Математика: 168\n\n                                                                                    Історія України: 151"

Оцінимо як виглядає 1 запис. Дуже схожі проблеми з полем Name. Спробуємо примінити уже готову функцію text_cleaner до цього поля.

In [46]:
grades <- text_cleaner(data$Grades[1])
grades

[1] "160 168 151 Українська мова: 160 Математика: 168 Історія України: 151"

In [50]:
lst <- str_split(grades, " ")

In [54]:
lst[1]

[[1]]
[1] "160"

Вийшло. Наступний етап розбиття інформації на частини. Нам потрібні лише 3 перші записи у цьому полі.

In [56]:
grades <- unlist(str_split(grades, " "))[1:3]
grades

[1] "160" "168" "151"

In [59]:
sapply(grades, text_cleaner) |> head()

160   168   151 
"160" "168" "151"

In [61]:
get_3_grades <- function(grades){
    unlist(strsplit(grades, " "))[1:3]
}

get_grades <- function(grades) {
    grades <- sapply(grades, text_cleaner)
    names(grades) <- NULL
    grades <- lapply(grades, get_3_grades)
    grades <- lapply(grades, as.numeric)
    grades
}  

get_grades(data$Grades) |> head()

[[1]]
[1] 160 168 151

[[2]]
[1] 194 200 200

[[3]]
[1] 200 185 200

[[4]]
[1] 194 172 176

[[5]]
[1] 172 192 170

[[6]]
[1] 174 185 170

In [62]:
grades <- get_grades(data$Grades)
grades |> head()

[[1]]
[1] 160 168 151

[[2]]
[1] 194 200 200

[[3]]
[1] 200 185 200

[[4]]
[1] 194 172 176

[[5]]
[1] 172 192 170

[[6]]
[1] 174 185 170

In [63]:
data <- data |>
    mutate(UkrGrades = unlist(lapply(grades, `[[`, 1)),
           MathGrades = unlist(lapply(grades, `[[`, 2)),
           HistoryGrades = unlist(lapply(grades, `[[`, 3)),
           Grades = NULL)
data |> head()

Rank Name              TotalGrade StudyType IsStudent UkrGrades MathGrades
1 1    120-7906910 В. О. 160.95     Бюджет    +         160       168       
2 2    Мініч К. В.       197.90     Бюджет    +         194       200       
3 3    Редколіс А. В.    194.00     Бюджет    +         200       185       
4 4    Скулкін О. В.     180.70     Контракт  +         194       172       
5 5    Мартинчук О. П.   179.50     Контракт  +         172       192       
6 6    Божок М. О.       177.40     Контракт  +         174       185       
  HistoryGrades
1 151          
2 200          
3 200          
4 176          
5 170          
6 170

#### 2.3.4. Трансформація поля `IsStudent`

Замінимо це поле на логічне та змінимо порядок стовпців:

In [64]:
data <- data |>
    mutate(IsStudent = ifelse(IsStudent == "+", T, F)) |>
    select(1:2, 6:8, 3:5)

In [65]:
data |> head(5)

Rank Name              UkrGrades MathGrades HistoryGrades TotalGrade
1 1    120-7906910 В. О. 160       168        151           160.95    
2 2    Мініч К. В.       194       200        200           197.90    
3 3    Редколіс А. В.    200       185        200           194.00    
4 4    Скулкін О. В.     194       172        176           180.70    
5 5    Мартинчук О. П.   172       192        170           179.50    
  StudyType IsStudent
1 Бюджет    TRUE     
2 Бюджет    TRUE     
3 Бюджет    TRUE     
4 Контракт  TRUE     
5 Контракт  TRUE

In [66]:
mean(data$UkrGrades, na.rm = T)

[1] 174.45

In [67]:
data |>
    group_by(IsStudent) |>
    summarise(UkrMean = mean(UkrGrades, na.rm = T),
             MathMean = mean(MathGrades, na.rm = T),
             HistoryMean = mean(HistoryGrades, na.rm = T))

IsStudent UkrMean  MathMean HistoryMean
1 FALSE     190.4000 178.4    182.0000   
2  TRUE     169.1333 161.0    159.5333

In [68]:
abit_parser <- function(url) {
    
    page <- read_html(url)    
    tables <- html_nodes(page, "table")    
    
    data <- as.data.frame(html_table(tables[1]), fill = T) |> select(-c(3,7))
    
    colnames(data) <- c("Rank", "Name", "TotalGrade", "StudyType", "Grades", "IsStudent")
    
    grades <- get_grades(data$Grades)
    
    data <- data |> 
        mutate(Name = unlist(clean_names(Name)),
               StudyType = ifelse(StudyType == "До наказу (б)", "Бюджет",
                                  ifelse(StudyType == "До наказу (к)", "Контракт",
                                         "Не навчається")),
               UkrGrades = unlist(lapply(grades, `[[`, 1)),
               MathGrades = unlist(lapply(grades, `[[`, 2)),
               HistoryGrades = unlist(lapply(grades, `[[`, 3)),
               Grades = NULL, 
               IsStudent = ifelse(IsStudent == "+", T, F)) |>
    select(1:2, 6:8, 3:5)
    
    data
}

In [69]:
data_new <- abit_parser("https://abit-poisk.org.ua/rate2022/direction/984185")

Warning message in lapply(grades, as.numeric):
"NAs introduced by coercion"


In [70]:
data_new 

Rank Name                 UkrGrades MathGrades HistoryGrades TotalGrade
1   1    120-7964783 В. В.    187       176        160           176.100   
2   2    Горецька Н. Л.       138       125        140           131.900   
3   3    Жарчинський М. С.    184       200        200           195.200   
4   4    Пильгун А. Ю.        184       200        186           192.400   
5   5    Ошейко В. О.         187       200        160           188.100   
6   6    Круш І. В.           177       192        180           185.100   
7   7    Строзюк Р. В.        174       192        180           184.200   
8   8    Лавренюк А. В.       194       185        162           183.100   
9   9    Дігалевич І. О.      172       192        176           182.800   
10  10   Сівачук М. О.        187       192        151           182.300   
11  11   Свирида А. Ю.        169       185        180           179.200   
12  12   Мельник Д. А.        190       172        173           177.600   
13  13   Бондар А. О.         190       168        176           176.200   
14  14   120-7996285 Б. М.    137       138        140           138.100   
15  15   Томусяк М. Б.        187       185        164           181.400   
16  16   Семенюк В. В.        172       176        176           174.800   
17  17   Давидюк Р. В.        182       168        167           172.000   
18  18   Тарасюк Д. О.        145       192        158           171.100   
19  19   Довгій Д. М.         154       185        158           170.300   
20  20   Носко Т. В.          157       176        160           167.100   
21  21   Сокальський О. І.    172       159        176           166.300   
22  22   Добровольський М. О. 160       168        156           163.200   
23  23   Літвінчук В. І.      151       176        146           162.500   
24  24   Шапран Н. В.         154       168        154           161.000   
25  25   Барковський А. О.    177       154        153           160.700   
26  26   Гнатюк І. І.         177       149        164           160.400   
27  27   Бурчак П. О.         145       165        164           158.800   
28  28   Дзем`юк В. Б.        149       159        170           158.200   
29  29   Гладуненко А. Р.     160       156        156           157.200   
30  30   Хажинський М. С.     169       149        156           156.400   
... ...  ...                  ...       ...        ...           ...       
162 161  Дімітрова А. О.      190       168        170           175.000   
163 162  Журба А. С.          180       176        164           174.800   
164 163  Білая Є. С.          184       172        167           174.600   
165 164  Старостюк В. С.      187       176        151           174.300   
166 165  Степовенко Ю. О.     172       165        200           174.100   
167 166  Сисюк В. В.          190       172        151           173.200   
168 167  Шелест О. О.         169       172        180           172.700   
169 168  Луцук В. Ю.          157       192        147           172.500   
170 169  Луцук В. Ю.          157       192        147           172.500   
171 170  Панько П. Т.         174       159        180           167.700   
172 171  Білий О. В.          154       176        167           167.600   
173 172  Шергей М. В.         169       168        160           166.700   
174 173  Грабовський П. І.    180       168        140           166.000   
175 174  Ткачук Д. В.         160       172        160           166.000   
176 175  Політько С. В.       180       159        160           165.500   
177 176  Будяківський М. Ю.   160       176        146           165.200   
178 177  Федас Р. В.          160       172        154           164.800   
179 178  Коханець В. І.       163       159        176           163.600   
180 179  Рудюк О. О.          169       159        162           162.600   
181 180  Криворучко А. М.     180       152        160           162.000   
182 181  Романюк О. В.        169       162        1